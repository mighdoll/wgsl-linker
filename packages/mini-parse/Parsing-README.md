**MiniParse** is a small TypeScript parser combinator library with an efficient regex based lexer.

## Parser Features

- Small code size makes **MiniParse** suitable for runtime embedding (< 5KB compressed).
- **MiniParse** is a combinator library.
  You write a grammar by combining simple TypeScript
  functions like `or()`, `repeat()`, and `seq()`.
  It's just TypeScript so it's easy to mix with your existing code,
  IDE, test frameworks, etc.
- **MiniParse** is a Parsing Expression Grammar (PEG) parser.
  It parses top down, using recursive descent with backtracking.
  Top down parsing is easy to understand.
- Parsers are modular - every grammar fragment is also a parser and can be tested and reused independently.
- Extensive debug tracing for use while developing grammars is built in.
  (Tracing is automatically removed from production builds).
- Error reporting is included, with full line context displayed.

## Advanced Features

- Swap lexers under parser control. This is handy for parsing quotes,
  or for mixed language parsing (e.g. html in jsx).
- Stack parsers to parse things that can appear almost anywhere in your grammar.
  Handy for things like nested comments, semantic comments (jsdoc), or annotations.
- Named accumulators make it easy to collect parsing results from deeply nested sub parsers.

## Parsing

Typically you can write a parser by mixing combinator functions from the library.

Here's a parser for a simple sum expression:

```ts
export const simpleSum = seq(num, or("+", "-"), num);
```

Here's a parser for nested block comments:

```ts
export const blockComment: Parser<void> = seq(
  "/*",
  repeat(or(() => blockComment, anyNot("*/"))),
  req("*/")
);
```

The example above uses the combinators:
`seq()`, `repeat()`, `or()`, `anyNot()`, and `req()`.
More combinators are available and documented in [ParserCombinator.ts](./src/ParserCombinator.ts)

To any combinator that accepts a parser as an argument, you can pass:
- another parser
- a function that returns a parser - uses the returned parser, but calls the function lazily
- a string - a parser that accepts any token exactly matching the string

## Lexer

Parsing relies on a lexer to divide the source string into parts, called tokens.
(You could produce a token per character, but that would be comparatively slow)

Here's an example token matcher to identify three kinds of tokens for parsing
simple arithmetic expressions.

```ts
export const simpleTokens = tokenMatcher({
  number: /\d+/,
  symbol: matchOneOf("( ) ^ + - * /"),
  ws: /\s+/,
});
```

To create a lexer that walks over a source text producing tokens, use:

```ts
const text = "3 + 4";
const lexer = matchingLexer(text, simpleTokens);
```

By default, the lexer will skip over whitespace tokens that you designate with the name 'ws'.
You can configure whether and which tokens to skip statically by passing the set of
names skip when you construct the `matchingLexer`, or dynamically while parsing
by using the `tokensIgnore()` combinator.

## Running the Parser

To run a parser, just give it a lexer attached to your source text.

```ts
const result = simpleSum.parse({ lexer });
```

The result will contain the combined results of the parsers, in this case `["3", "+", "4"]`.

### Selecting Parsing Results
Typically, it's convenient to use `.map()` to select the relevant parts from a successful parse 
and do a bit of format conversion. 

This parser will return a number rather than a string:
```
const int = num.map((r) => parseInt(r.value));
```

Here's an example that even does some computation, and returns a numeric sum or difference
of the whole expression.
It parses the same text as `simpleSum` above, but converts to numbers and then adds or subtracts.
```
// return the numeric sum, rather than a sequence of strings
export const sumResults = seq(int, or("+", "-"), int).map((r) => {
  const [a, op, b] = r.value;
  return op === "+" ? a + b : a - b;
});
```

Note that `.map()` is only called on successful parses of the mapped expression, 
if the expression fails, the parser will backtrack and try any alternatives in the grammar 
and `.map()` will not be called on the failed part of the parse.

### App State 
For larger parsers, you'll typically convert the parsed text into an intermediate form, sometimes
called an [Abstract Syntax Tree](https://en.wikipedia.org/wiki/Abstract_syntax_tree).

`.parse()` allows you to pass an application specific data structure that's visible in `.map()` 
for convenience in building the abstract syntax tree with each successfully parsed element.

```ts
type ASTElem = BinOpElem;

interface BinOpElem {
  kind: "binOp";
  left: number | BinOpElem;
  right: number | BinOpElem;
  op: "+" | "-";
}

export const sumElem = seq(int, or("+", "-"), int).map((r) => {
  const [a, op, b] = r.value;
  const binOpElem:BinOpElem = {
    kind: "binOp",
    left: a,
    right: b,
    op: op as "+" | "-",
  };
  r.app.state.push(binOpElem);
});

```

### Named Results

For complicated nested or repeating parsers,
it's convenient to assign names to particular results of interest.
You can use a `Symbol` or a string to name a result using the `.named()` 
method on all parsers.
Multiple results with the same name are accumulated into an array.

```ts
const op = or("+", "-");

export const namedSum = seq(
  int,
  repeat(seq(op, int).named("opRights")) // accumulate an array of [op, int] pairs
).map((r) => {
  const { opRights } = r.named;
  const left = r.value[0];
  if (!opRights) return left;
  return opRights.reduce((acc, opRight) => {
    const [op, right] = opRight;
    return op === "+" ? acc + right : acc - right;
  }, left);
});
```

## Debug Tracing
For debugging your grammar, it's useful to debug your grammar in pieces.
One of the nice features of parser combinators is that every part of the grammar is
independently testable.

To print out the progress of parsing: 
1. Call `enableTracing()` to turn on the tracing facility (normally off and removed from prod builds)
1. Call `.trace(opts?)` on any Parser. See `TraceOptions` for options controlling trace levels.
1. Add application relevant trace names to any parser using `.traceName()` or `setTraceName()`.
  * Use `.traceName()` on any parser to set the trace name for debugging.
  * Alternately, you can use `setTraceName()` protected by a `tracing`
    global and the javascript bundler will remove the code in production builds to save a few bytes.

  ```ts
  if (tracing) {
    const names: Record<string, Parser<unknown>> = {
      fnCall,
      fnParam,
      fnParamList,
      fnDecl,
    };

    Object.entries(names).forEach(([name, parser]) => {
      setTraceName(parser, name);
    });
  ```

## Examples

[Calculator]() - classic PEG style parser in **MiniParse** syntax.

[Calculator with Results]() calculator example parser with inline calculation of results.

[WGSL-D]() parsing some of the WebGPU [WGSL](https://www.w3.org/TR/WGSL/#grammar-recursive-descent) shader language with `#import` and `#export` extensions.


## Special Situations

### tokens() combinator
Sometimes it's nice to let the grammar choose a different tokenizer 
to parse different sections of the source text. 
For example, to parse a programming language with quotes, 
you'll probably want a different tokenizer for the text inside of quotes:

```ts
const quote = seq('"', tokens(quoteTokens, repeat(nonQuote)), '"')
```

### .toParser()

In unusual cases, it can be handy to choose the next parser based
on information outside the grammar itself.
For example, to handle an `#ifdef` style preprocessor, the parsing
proceeds differently depending on whether the `#ifdef` is true or false,
and information outside the source text may be required to decide.
If you want to check a runtime dictionary to decide which parser to
use for the next tokens, than `.toParser` is of use.

### preParse() combinator

If the language you're parsing has some elements that can appear almost anywhere,
it'd be awkward to mention those elements at every possible position in the grammar.
Examples include nested block comments, comments containing semantic info, etc.

To handle pervasive elements, **MiniParse** offers an unusual feature called preparsing
that allows you to stack parsers. First the pre-parser will run, 
and if it fails to match at the current position, then the main parser will run.

```ts
const p = preParse(blockComments, mainParser);
```

Multiple preparsers can be attached. Preparsing can also be temporarily disabled
in the grammar, e.g. to disable comment skipping inside quotes.

Save preparsing for special situations.
If the pervasive elements are easy to find and can be skipped, 
then adding a few token types to skip in the lexer is simpler and faster.
That's typically the approach for white space.  

### app.context

There are two application specific objects that are passed to every parser: 
`state` and `context`. 
`app.state`, as mentioned above, is handy for accumulating application results of successful parses.

`app.context` is useful to store ephemeral application state discovered
during parsing. 
Like `app.state`, `app.context` is just for applications - **MiniParse** doesn't use it
and applications can read and write it using the `.map()` method on any parser.
But unlike `app.state` **MiniParse** will reset `app.context` when a sub-parser fails and backtracks. 
`app.context` is passed to child parsers, but doesn't accumulate to parent parsers.

An example of using `app.context` is for parsing nested `#ifdef` `#endif` clauses. 
`app.context` is a good place to store the stack of active/inactive states discovered while
parsing. 

### Left recursion

Left recursive rules are typically disallowed in top down parsers, including MiniParse.
In the parser combinator setting, it's obvious why - a function calling itself
in its first statement is going to recurse forever.
Best to write the grammar so that recursion is in the middle or at the end.
See the block comment example or the calculator example.


## Future Work

PEG parsers like MiniParse can be sped up using a memoization algorithm called packrat parsing.

[Tratt](https://tratt.net/laurie/research/pubs/html/tratt__direct_left_recursive_parsing_expression_grammars/)
describes a technique to allow some left recursive rules, based on
[Warth](https://tinlizzie.org/VPRIPapers/tr2007002_packrat.pdf)'s proposal for left recursion
with packrat parsing.
[Rossum](https://medium.com/@gvanrossum_83706/left-recursive-peg-grammars-65dab3c580e1) also
has pursued this approach for Python.
But per Tratt, note that the resulting parse order is not as predictable, and there
are issues with rules that are simultaneously left and right recursive.

Allowing a regex as a parser argument would be convenient to avoid the need for a separate lexer in some cases.

## Choosing a Parsing Approach

Is **MiniParse** right for your project? Consider the alternatives:

* **Full Custom Parser** - _maximum speed and ultimate malleability, lots of work._

  For maximum speed and control, write a dedicated parser directly in Typescript.
  This is the most effort, but if you're writing a production compiler and need to squeeze
  every millisecond, it's worth it.
  Otherwise use a parser generator tool suite or a parser combinator library.

* **Parser Generator** - _high speed, some work to adopt._

  Parser generators statically analyze and precompile a grammar description language.
  These mature tools can be a bit big, but there's lots of documentation,
  rich ecosystems of example code and support tools.

  Worthy examples include:
  [Nearley](https://nearley.js.org/),
  [Lezer](https://lezer.codemirror.net/),
  [Antlr](https://www.antlr.org/),
  or perhaps [Ohm](https://ohmjs.org/).
  Each parser generator has its own textual format to describe the grammar. The library
  compiles into an execution format before parsing.
  Each of the above libraries uses a different base algorithm (Earley, GLR, LL, Packrat)
  with different tradeoffs, but all have evolved robust features to classic parsing
  problems of error recovery, left recursion, producing parse results, tracing, etc.

  Parser generators are typically more complicated to adopt than parser combinator libraries,
  less flexible, require a separate build step, and they're more code than typical parser combinators.
  But for demanding parsing jobs, the complexity of a parser generator tool is
  easily worth the investment.

* **Parser Combinators** - _lower speed, most flexibility, lightweight adoption._

  Parser combinators define a grammar by mixing simple TypeScript functions
  provided by the library or written by the user (aka combinator functions).
  Execution of the grammar involves simply running these functions.
  The simplicity makes parser combinators flexible and easy to adopt - you're using
  TypesScript for everything.

  Parser combinators are interpreting rather than compiling the grammar in advance,
  so they're slower to run. But they're plenty fast enough for most purposes.

  In the Parser Combinator category, **MiniParse** has a few interesting features
  and is notably lightweight.
  As you're parser shopping, also consider other worthy and more mature libraries
  in the TypeScript parser combinator category like:
  [ts-parsec](https://github.com/microsoft/ts-parsec)
  and [ParJS](https://github.com/GregRos/parjs).
