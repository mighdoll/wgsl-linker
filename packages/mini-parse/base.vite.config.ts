import { UserConfig } from "vite";
import tsconfigPaths from "vite-tsconfig-paths";
import dts from "vite-plugin-dts";
import { resolve } from "path";

export function baseViteConfig(): UserConfig {
  return {
    plugins: [
      tsconfigPaths(),
      dts(), // generate .d.ts files
    ],
    build: {
      lib: {
        name: "mini-parse",
        entry: [resolve(__dirname, "src/index.ts")],
      },
      minify: false,
      sourcemap: true,
    },
  };
}
