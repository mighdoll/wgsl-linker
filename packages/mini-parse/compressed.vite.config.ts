/// <reference types="vitest" />
import { LibraryOptions, defineConfig } from "vite";
// import { visualizer } from "rollup-plugin-visualizer";
import { baseViteConfig } from "./vite-base-config.js";

// Note that this will include the debug tracing code, so
// so the size estimate is an overestimate of production size

const config = baseViteConfig();
config.build!.emptyOutDir = false;
config.build!.minify = "terser";
(config.build!.lib as LibraryOptions).formats = ["es", "cjs"];
(config.build!.lib as LibraryOptions).fileName = "minified";

// config.plugins?.push(visualizer({ brotliSize: true, gzipSize: true })); // generate stats.html size report

export default defineConfig(config);
