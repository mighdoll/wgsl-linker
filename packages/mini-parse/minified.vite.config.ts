/// <reference types="vitest" />
import { LibraryOptions, defineConfig } from "vite";
import { baseViteConfig } from "./base.vite.config.js";
// import { visualizer } from "rollup-plugin-visualizer";

// Note that this will include the debug tracing code, so
// so the size estimate is an overestimate of production size

const config = baseViteConfig();
config.build!.emptyOutDir = false;
config.build!.minify = "terser";
(config.build!.lib as LibraryOptions).formats = ["es", "cjs"];
(config.build!.lib as LibraryOptions).fileName = "minified";

// config.plugins?.push(visualizer({ brotliSize: true, gzipSize: true })); // generate stats.html size report

export default defineConfig(config);
