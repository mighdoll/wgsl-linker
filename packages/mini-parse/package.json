{
  "name": "mini-parse",
  "version": "0.2.0",
  "files": [
    "dist",
    "src"
  ],
  "scripts": {
    "pr": "printenv",
    "build": "run-s build:main build:util build:compressed",
    "build:main": "vite build",
    "build:util": "vite -c test-util.vite.config.ts build",
    "build:compressed": "vite -c compressed.vite.config.ts build",
    "build:brotli": "brotli dist/minified.cjs",
    "test": "vitest",
    "organize": "organize-imports-cli tsconfig.json",
    "format": "prettier-eslint \"src/**/*.ts\" --write",
    "lint": "eslint src"
  },
  "type": "module",
  "repository": "github:mighdoll/wgsl-linker",
  "exports": {
    ".": {
      "import": "./dist/index.js",
      "require": "./dist/index.cjs",
      "types": "./dist/index.d.ts"
    },
    "./minified": {
      "import": "./dist/minified.js",
      "require": "./dist/minified.cjs",
      "types": "./dist/index.d.ts"
    },
    "./test-util": {
      "import": "./dist/testUtil.js",
      "types": "./dist/test-util/testUtil.d.ts"
    }
  },
  "devDependencies": {
    "@types/node": "^20.11.16",
    "@typescript-eslint/eslint-plugin": "^6.21.0",
    "@typescript-eslint/parser": "^6.21.0",
    "@vitest/ui": "^1.1.1",
    "berry-pretty": "^0.0.4",
    "brotli": "^1.3.3",
    "esbuild": "^0.20.0",
    "eslint": "^8.56.0",
    "eslint-config-prettier": "^9.1.0",
    "npm-run-all": "^4.1.5",
    "organize-imports-cli": "^0.10.0",
    "prettier-eslint-cli": "^8.0.1",
    "rollup-plugin-visualizer": "^5.12.0",
    "terser": "^5.27.0",
    "typescript": "^5.3.3",
    "vite": "^5.0.12",
    "vite-plugin-dts": "^3.6.4",
    "vite-tsconfig-paths": "^4.2.1",
    "vitest": "^1.2.2"
  }
}
