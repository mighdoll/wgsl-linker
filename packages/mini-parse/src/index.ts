export * from "./MatchingLexer.js";
export * from "./Parser.js";
export * from "./ParserCombinator.js";
export * from "./ParserTracing.js";
export * from "./TokenMatcher.js";
export * from "./ParserLogging.js";
export * from "./test/TestParse.js";
export * from "./test/LogCatcher.js";
