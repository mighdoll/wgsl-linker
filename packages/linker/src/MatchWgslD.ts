import { matchOneOf, tokenMatcher } from "mini-parse";

/** token matchers for wgsl with #directives */

export const eol = /\n/;
export const directive = /#[a-zA-Z_]\w*/;
export const notDirective = /[^#\n]+/;

const symbolSet =
  "& && -> @ / ! [ ] { } : , = == != > >= < << <= % - -- " + // '>>' elided for template parsing, e.g. vec2<vec2<u8>>
  ". + ++ | || ( ) ; * ~ ^ // /* */ += -= *= /= %= &= |= ^= >>= <<= <<";
const symbol = matchOneOf(symbolSet);

/** matching tokens at wgsl root level */
export const mainTokens = tokenMatcher(
  {
    directive,
    attr: /@[a-zA-Z_]\w*/,
    word: /[a-zA-Z_]\w*/, // LATER consider making this 'ident' per wgsl spec (incl. non-ascii)   word,
    digits: /(?:0x)?[\d.]+[iuf]?/, // LATER parse more wgsl number variants
    symbol,
    ws: /\s+/,
  },
  "main"
);

export const moduleTokens = tokenMatcher(
  {
    ws: /\s+/,
    moduleName: /[a-zA-Z_][\w./-]*/,
  },
  "moduleName"
);

/** matching tokens at the start of a '//' line comment that might contain #directives */
export const lineCommentTokens = tokenMatcher(
  {
    directive,
    ws: /[ \t]+/, // note ws must be before notDirective
    notDirective,
    eol,
  },
  "lineComment"
);

/** matching tokens while parsing directive parameters #export foo(param1, param2) */
export const argsTokens = tokenMatcher(
  {
    directive,
    relPath: /[.][/\w._-]+/,
    arg: /[\w._-]+/,
    symbol,
    ws: /[ \t]+/, // don't include \n, so we can find eol separately
    eol,
  },
  "argsTokens"
);
