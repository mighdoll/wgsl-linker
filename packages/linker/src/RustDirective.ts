import {
  Parser,
  TagRecord,
  withTags,
  kind,
  opt,
  or,
  repeat,
  seq,
  setTraceName,
  text,
  tokens,
  tracing,
  withSep,
} from "mini-parse";
import {
  ImportTree,
  SegmentList,
  SimpleSegment,
  Wildcard,
} from "./ImportTree.js";
import { treeImportTokens } from "./MatchWgslD.js";
import { makeElem } from "./ParseSupport.js";

const word = kind(treeImportTokens.word);

// forward reference (for mutual recursion)
let importTree: Parser<any, any> = null as any;

const simpleSegment = withTags(
  seq(word.tag("segment"), opt(seq("as", word.tag("as")))).map(
    (r) => new SimpleSegment(r.tags.segment[0], r.tags.as?.[0])
  )
);

const wildCard = text("*").map(() => Wildcard._);

const segmentList = withTags(
  seq("{", withSep(",", () => or(importTree, wildCard)).tag("list"), "}").map(
    (r) => {
      return new SegmentList(r.tags.list.flat());
    }
  )
);

const pathSegment = or(simpleSegment, wildCard, segmentList);

const pathExtension = withTags(
  repeat(seq("::", pathSegment.tag("segments"))).map((r) => {
    return r.tags.segments || [];
  })
);

importTree = seq(simpleSegment, pathExtension).map((r) => {
  return new ImportTree(r.value.flat());
});

/** parse a Rust style wgsl import statement.
 * The syntax is like 'use' in Rust.
 * 'self' references are not currenlty supported. */
export const rustImport = tokens(
  treeImportTokens,
  seq(
    or("#import", "import"), // bevy uses #import, but std says import
    importTree.tag("imports"),
    opt(";")
  ).map((r) => {
    const e = makeElem("treeImport", r, ["imports", "from"]);

    r.app.state.push(e);
  })
);

// enableTracing();
if (tracing) {
  const names: Record<string, Parser<unknown, TagRecord>> = {
    segmentList,
    simpleSegment,
    wildCard,
    pathSegment,
    pathExtension,
    importTree,
    rustImport,
  };

  Object.entries(names).forEach(([name, parser]) => {
    setTraceName(parser, name);
  });
}
