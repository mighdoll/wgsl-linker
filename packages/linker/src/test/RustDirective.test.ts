import { tokens } from "mini-parse";
import { expect, test } from "vitest";
import { argsTokens } from "../MatchWgslD.js";
import { rustImport } from "../RustDirective.js";
import { testAppParse } from "./TestUtil.js";
import { dlogOpt } from "berry-pretty";

const testRustImport = tokens(argsTokens, rustImport);

test("import a::b::c", (ctx) => {
  const { appState } = testAppParse(testRustImport, ctx.task.name);
  expect(appState).toMatchSnapshot();
});

test("import my::lighting::pbr;", (ctx) => {
  const { appState } = testAppParse(testRustImport, ctx.task.name);
  expect(appState).toMatchSnapshot();
});

test("import my::lighting::{ pbr };", (ctx) => {
  const { appState } = testAppParse(testRustImport, ctx.task.name);
  expect(appState).toMatchSnapshot();
});

test(`import my::lighting::{ pbr, jelly };`, (ctx) => {
  const { appState } = testAppParse(testRustImport, ctx.task.name);
  expect(appState).toMatchSnapshot();
});

test("import my::lighting::{ pbr, jelly::jam };", (ctx) => {
  const { appState } = testAppParse(testRustImport, ctx.task.name);
  expect(appState).toMatchSnapshot();
});

test("import my::lighting::{ pbr as lights }", (ctx) => {
  const { appState } = testAppParse(testRustImport, ctx.task.name);
  expect(appState).toMatchSnapshot();
});

test("import a::*", (ctx) => {
  const { appState } = testAppParse(testRustImport, ctx.task.name);
  expect(appState).toMatchSnapshot();
});

test("import a::{b, c}::*", (ctx) => {
  const { appState } = testAppParse(testRustImport, ctx.task.name);
  expect(appState).toMatchSnapshot();
});

test("multiline import", () => {
  const src = `import a::
        {b, c}::*;`;
  const { appState } = testAppParse(testRustImport, src);
  expect(appState).toMatchSnapshot();
});
