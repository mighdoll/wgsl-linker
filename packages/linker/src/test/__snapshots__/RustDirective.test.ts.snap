// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`import a::* 1`] = `
[
  {
    "end": 11,
    "imports": ImportTree {
      "segments": [
        SimpleSegment {
          "as": undefined,
          "name": "a",
        },
        Wildcard {
          "wildcard": "*",
        },
      ],
    },
    "kind": "treeImport",
    "start": 0,
  },
]
`;

exports[`import a::{b, c}::* 1`] = `
[
  {
    "end": 19,
    "imports": ImportTree {
      "segments": [
        SimpleSegment {
          "as": undefined,
          "name": "a",
        },
        SegmentList {
          "list": [
            ImportTree {
              "segments": [
                SimpleSegment {
                  "as": undefined,
                  "name": "b",
                },
              ],
            },
            ImportTree {
              "segments": [
                SimpleSegment {
                  "as": undefined,
                  "name": "c",
                },
              ],
            },
          ],
        },
        Wildcard {
          "wildcard": "*",
        },
      ],
    },
    "kind": "treeImport",
    "start": 0,
  },
]
`;

exports[`import a::b::c 1`] = `
[
  {
    "end": 14,
    "imports": ImportTree {
      "segments": [
        SimpleSegment {
          "as": undefined,
          "name": "a",
        },
        SimpleSegment {
          "as": undefined,
          "name": "b",
        },
        SimpleSegment {
          "as": undefined,
          "name": "c",
        },
      ],
    },
    "kind": "treeImport",
    "start": 0,
  },
]
`;

exports[`import my::lighting::{ pbr }; 1`] = `
[
  {
    "end": 29,
    "imports": ImportTree {
      "segments": [
        SimpleSegment {
          "as": undefined,
          "name": "my",
        },
        SimpleSegment {
          "as": undefined,
          "name": "lighting",
        },
        SegmentList {
          "list": [
            ImportTree {
              "segments": [
                SimpleSegment {
                  "as": undefined,
                  "name": "pbr",
                },
              ],
            },
          ],
        },
      ],
    },
    "kind": "treeImport",
    "start": 0,
  },
]
`;

exports[`import my::lighting::{ pbr as lights } 1`] = `
[
  {
    "end": 38,
    "imports": ImportTree {
      "segments": [
        SimpleSegment {
          "as": undefined,
          "name": "my",
        },
        SimpleSegment {
          "as": undefined,
          "name": "lighting",
        },
        SegmentList {
          "list": [
            ImportTree {
              "segments": [
                SimpleSegment {
                  "as": "lights",
                  "name": "pbr",
                },
              ],
            },
          ],
        },
      ],
    },
    "kind": "treeImport",
    "start": 0,
  },
]
`;

exports[`import my::lighting::{ pbr, jelly }; 1`] = `
[
  {
    "end": 36,
    "imports": ImportTree {
      "segments": [
        SimpleSegment {
          "as": undefined,
          "name": "my",
        },
        SimpleSegment {
          "as": undefined,
          "name": "lighting",
        },
        SegmentList {
          "list": [
            ImportTree {
              "segments": [
                SimpleSegment {
                  "as": undefined,
                  "name": "pbr",
                },
              ],
            },
            ImportTree {
              "segments": [
                SimpleSegment {
                  "as": undefined,
                  "name": "jelly",
                },
              ],
            },
          ],
        },
      ],
    },
    "kind": "treeImport",
    "start": 0,
  },
]
`;

exports[`import my::lighting::{ pbr, jelly::jam }; 1`] = `
[
  {
    "end": 41,
    "imports": ImportTree {
      "segments": [
        SimpleSegment {
          "as": undefined,
          "name": "my",
        },
        SimpleSegment {
          "as": undefined,
          "name": "lighting",
        },
        SegmentList {
          "list": [
            ImportTree {
              "segments": [
                SimpleSegment {
                  "as": undefined,
                  "name": "pbr",
                },
              ],
            },
            ImportTree {
              "segments": [
                SimpleSegment {
                  "as": undefined,
                  "name": "jelly",
                },
                SimpleSegment {
                  "as": undefined,
                  "name": "jam",
                },
              ],
            },
          ],
        },
      ],
    },
    "kind": "treeImport",
    "start": 0,
  },
]
`;

exports[`import my::lighting::pbr; 1`] = `
[
  {
    "end": 25,
    "imports": ImportTree {
      "segments": [
        SimpleSegment {
          "as": undefined,
          "name": "my",
        },
        SimpleSegment {
          "as": undefined,
          "name": "lighting",
        },
        SimpleSegment {
          "as": undefined,
          "name": "pbr",
        },
      ],
    },
    "kind": "treeImport",
    "start": 0,
  },
]
`;

exports[`multiline import 1`] = `
[
  {
    "end": 29,
    "imports": ImportTree {
      "segments": [
        SimpleSegment {
          "as": undefined,
          "name": "a",
        },
        SegmentList {
          "list": [
            ImportTree {
              "segments": [
                SimpleSegment {
                  "as": undefined,
                  "name": "b",
                },
              ],
            },
            ImportTree {
              "segments": [
                SimpleSegment {
                  "as": undefined,
                  "name": "c",
                },
              ],
            },
          ],
        },
        Wildcard {
          "wildcard": "*",
        },
      ],
    },
    "kind": "treeImport",
    "start": 0,
  },
]
`;
